File: src/com/train/algorithm/list/implementInC++/ReverseLinkedList.cpp
//
// Created by 罗斯 on 12/8/2022.
//
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* cur = head;
        ListNode* next;

        while(cur){
            next = cur ->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }
        return prev;
    }
};

File: src/com/train/algorithm/list/implementInPython/ReverseLinkedList.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev, curr, next = None, head, None
        while curr:
            next, curr.next = curr.next, prev
            prev, curr = curr, next

        return prev